File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/api/node/NodeTypeRegistry.java
package de.cubbossa.pathfinder.api.node;

import de.cubbossa.pathfinder.api.misc.NamespacedKey;
import java.util.Collection;

public interface NodeTypeRegistry {
  <N extends Node<N>> NodeType<N> getType(NamespacedKey key);

  Collection<NamespacedKey> getTypeKeys();

  Collection<NodeType<? extends Node<?>>> getTypes();

  <N extends Node<N>> void register(NodeType<N> type);

  <N extends Node<N>> void unregister(NodeType<N> type);

  void unregister(NamespacedKey key);
}

File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/node/NodeTypeRegistry.kt
package de.cubbossa.pathfinder.node

import de.cubbossa.disposables.Disposable
import de.cubbossa.pathfinder.misc.NamespacedKey

interface NodeTypeRegistry : Disposable {

    val typeKeys: Collection<NamespacedKey>
    val types: Collection<NodeType<*>>

    fun <N : Node> getType(key: NamespacedKey): NodeType<N>?

    fun <N : Node> register(type: NodeType<N>)

    fun <N : Node> unregister(type: NodeType<N>)

    fun unregister(key: NamespacedKey)
}
